src/libraries/System.Private.CoreLib/src/System/IO/FileStream.Windows.cs

```diff
@@ -389,6 +389,7 @@ private void SetLengthCore(long value)
                 SeekCore(_fileHandle, value, SeekOrigin.Begin);
             if (!Interop.Kernel32.SetEndOfFile(_fileHandle))
             {
+                SeekCore(_fileHandle, origPos, SeekOrigin.Begin);
```

Comment:
Have you looked at the SetLengthCore implementation on Unix in FileStream.Unix.cs?  Seems like we have a similar issue there.

Reply:
On Unix, `SetLength` doesn't throw if it exceeds the available space.
However, `FileStream.Write` does throw and does not restore `Position` when that happens, which causes that the reported error occurs in a similar fashion. 

I think we can fix that case (restore Position when `Write` fails) in this PR as well.
https://github.com/dotnet/runtime/blob/3fca0837b821509ef5e0130faa265b6e3e556e2a/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.Unix.cs#L666-L672

Reply:
> On Unix, SetLength doesn't throw if it exceeds the available space.

It does potentially throw if ftruncate fails with an error:
https://github.com/dotnet/runtime/blob/e5f6b4417c3a503fe83dccd061a21d54b61a89fa/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.Unix.cs#L413
It's not 100% clear to me what conditions could trigger that to fail, but the man page says it can fail with EIO if "An I/O error occurred while reading from or writing to a file system."

Reply:
I tried to find a way to make `ftruncate` fail since it cannot fail with ENOSPC (as windows' `SetEndOfFile` similarly does) but I didn't succeed.
Obviously the common errors like EINVAL cannot occur because SetLength validates up-front the FileStream instance and the arguments.

On my effort I tried to get a EPERM error by setting `ulimit -f 100` and then calling `ftruncate(fileno(fd), (__off64_t)60000)`, but that didn't work.

I am convinced that there are ways in which the ftruncate function may fail when called from .NET, I just don't find the scenario to fix and test what you are originally asking for.

Any suggestion is appreciated.

Reply:
David and I looked at this offline, we found a method called [`posix_fallocate`](https://www.man7.org/linux/man-pages/man3/posix_fallocate.3.html) which seems to do the same as `ftruncate`, but can return `ENOSPC`.

I noticed [we do check if we have it in mono](https://github.com/dotnet/runtime/search?q=posix_fallocate), but I did not find any usage instances. I found neither checks nor usage instances in libraries.

@jkotas do you have any additional insight regarding changing the `ftruncate` method for `posix_fallocate`?